keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1000 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+
+Contextual logging replaces the global logger by passing a `logr.Logger`
+instance into functions via a `context.Context` or an explicit
+parameter, building on top of [structured
+logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging).
+
+It enables the caller to:
+- attach key/value pairs that get included in all log messages
+- add names that describe which component or operation triggered a log messages
+- reduce the amount of log messages emitted by a callee by changing the
+  verbosity
+
+This works without having to pass any additional information into the callees
+because the additional information or settings are stored in the modified
+logger instance that is passed to it.
+
+Third-party components that use Kubernetes packages like client-go are no
+longer forced to use klog. They can choose an arbitrary implementation of
+`logr.Logger` and configure it as desired.
+
+During unit testing, each test case can use its own logger to ensure that log
+output is associated with the test case.
+
+## Motivation
+
+### Goals
+
+- Remove `k8s.io/klog` imports and usage from all packages.
+- Grant the caller of a function control over logging inside that function.
```

Comment:
I think I don't understand the justification of this (per function caller control) vs package level control or static initialization control.  I see https://github.com/kubernetes/enhancements/pull/3078/files#diff-b3356a3a6ca8f5573efb34e6ff97163a3e2778fcc8c48b02b984f7bfb8c88c7bR951, but that doesn't really make the case that the chosen approach is actually better for most users.

I'm a bit concerned about the scope of complexity adding to logging here - while I acknowledge there are many problems we *could* solve, the value of this for client-go seems vastly less than the cost and complexity described here.

Reply:
For instance, the alternative of "use some other generic log library in client-go and set up some import blocks so that you can import klog" would seem to accomplish the described gained value here.

Reply:
Let's add ", either by passing a logger into the function or by configuring the object that a method belongs to."

Global per-package loggers still have the problem that their output cannot be captured properly during parallel unit tests.